### ISSUE TITLE
[BottomSheetBehavior] Allow programmatically setting `STATE_HIDDEN` even when non-hideable

### ISSUE THREAD

**Issue by OxygenCobalt** [material-components/material-components-android#2786]
**Is your feature request related to a problem? Please describe.**
My app has a Bottom Sheet that I do not want to be hidden by the user, as it contains important information. However, it still needs to be hidden programmatically when there is nothing to show. However, if I disable hiding with `isHideable`, I can no longer change the state programmatically due to [these lines in `BottomSheetBehavior`](https://github.com/material-components/material-components-android/blob/cd37b9170c149de80ca5729035b5a2ad33afc9dd/lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java#L1227):

```
if (!hideable && state == STATE_HIDDEN) {
  Log.w(TAG, "Cannot set state: " + state);
  return;
}
```

**Describe the solution you'd like**
The removal of the `isHideable` check when setting the state to `STATE_HIDDEN`. Allow it to be set programmatically, and change `isHideable` to only apply to user input.

**Describe alternatives you've considered**
I could modify `isHideable` during run-time as such:
```kotlin
if (hide) {
    behavior.isHideable = true
    behavior.state = BottomSheetBehavior.STATE_HIDDEN
} else {
    behavior.isHideable = false
    behavior.state = BottomSheetBehavior.STATE_COLLAPSED
}
```

But I worry about the implications this might have for state consistency.

**Comment by drchen** [comment]
As for the state consistency - the alternative you mentioned is actually better, in my personal opinion. 

I'll close the issue for now since I don't think it's a good idea to alter a released behavior globally when there's a working alternative which makes sense logically.

**Comment by OxygenCobalt** [comment]
Okay, no problem.

**Comment by OxygenCobalt** [comment]
Sorry to revive this, but I've actually encountered layout issues with the workaround @drchen. For some reason setting those flags to true early on enough in the layout process causes the bottom sheet to flicker to the expanded state and then animate back to the collapsed state. Sadly, I can't seem to reproduce this in a test app, so I can't file any meaningful issue or debug it. All I know is that not mutating `isHideable` and just keeping it at `true` fixes the issue.

Would it be reasonable to add a private overrideable function (kind of like `shouldSkipHalfExpandedWhenDragging`) that allows me to control whether hiding *gestures* are enabled while still allowing me to use the hideable state itself? This would not break compatibility since the actual semantics of `isHideable` are not changed. I also just don't want to vendor BottomSheetBehavior so that I can disable hide gestures myself, which I will need to in order to fix the issue.

**Comment by drchen** [comment]
The issue you mentioned sounds a bit bizarre to me since setting hideable to true basically does nothing except for setting the variable and update accessibility actions. 

The workaround you proposed makes sense to me. However we don't really have people right now to work on it. If you can create a PR for this I would love to review it. : )

**Comment by OxygenCobalt** [comment]
> The issue you mentioned sounds a bit bizarre to me since setting hideable to true basically does nothing except for setting the variable and update accessibility actions.

Yeah, looking at the code I'm starting to think it's a race condition stemming from:
1. The amount of layouts I have to do regarding the bottom sheet (I have to move UI elements around, update window inset padding, etc), which causes layout-time to extend into runtime.
2. I hide and unhide the sheet in a function observing a `StateFlow`, which might cause some co-routine shenanigans that allows the layout and the hiding code to occur simultaniously.

The end result is some race condition that I think results in the following logic:
- I set `isHideable` to false
- The layout code runs, finds that `isHideable` is false, and then does *something* that results in the sheet being laid out as expanded
- `setHideable` updates the state to COLLAPSED, causing it to collapse back down

> The workaround you proposed makes sense to me. However we don't really have people right now to work on it. If you can create a PR for this I would love to review it. : )

Okay, I've already implemented it myself, so I can just PR it.
